File: vane-regions/src/main/java/org/oddlama/vane/regions/RegionBlueMapLayer.java
package org.oddlama.vane.regions;

import java.util.UUID;
import org.oddlama.vane.annotation.config.ConfigBoolean;
import org.oddlama.vane.annotation.config.ConfigDouble;
import org.oddlama.vane.annotation.config.ConfigInt;
import org.oddlama.vane.annotation.lang.LangMessage;
import org.oddlama.vane.core.lang.TranslatedMessage;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.core.module.ModuleComponent;
import org.oddlama.vane.regions.region.Region;

public class RegionBlueMapLayer extends ModuleComponent<Regions> {

    @ConfigBoolean(def = false, desc = "If the marker set should be hidden by default.")
    public boolean configHideByDefault;

    @ConfigBoolean(
        def = true,
        desc = "Set to false to make the area markers visible through terrain and other objects."
    )
    public boolean configDepthTest;

    @ConfigInt(def = 2, min = 1, desc = "Area marker line width.")
    public int configLineWidth;

    @ConfigInt(def = 0xffb422, min = 0, max = 0xffffff, desc = "Area marker fill color (0xRRGGBB).")
    public int configFillColor;

    @ConfigDouble(def = 0.1, min = 0.0, max = 1.0, desc = "Area marker fill opacity.")
    public double configFillOpacity;

    @ConfigInt(def = 0xffb422, min = 0, max = 0xffffff, desc = "Area marker line color (0xRRGGBB).")
    public int configLineColor;

    @ConfigDouble(def = 1.0, min = 0.0, max = 1.0, desc = "Area marker line opacity.")
    public double configLineOpacity;

    @LangMessage
    public TranslatedMessage langLayerLabel;

    @LangMessage
    public TranslatedMessage langMarkerLabel;

    private RegionBlueMapLayerDelegate delegate = null;

    public RegionBlueMapLayer(final Context<Regions> context) {
        super(context.group("BlueMap", "Enable BlueMap integration."));
    }

    public void delayedOnEnable() {
        final var plugin = getModule().getServer().getPluginManager().getPlugin("BlueMap");
        if (plugin == null) {
            return;
        }

        delegate = new RegionBlueMapLayerDelegate(this);
        delegate.onEnable(plugin);
    }

    @Override
    public void onEnable() {
        scheduleNextTick(this::delayedOnEnable);
    }

    @Override
    public void onDisable() {
        if (delegate != null) {
            delegate.onDisable();
            delegate = null;
        }
    }

    public void updateMarker(final Region region) {
        if (delegate != null) {
            delegate.updateMarker(region);
        }
    }

    public void removeMarker(final UUID regionId) {
        if (delegate != null) {
            delegate.removeMarker(regionId);
        }
    }

    public void updateAllMarkers() {
        if (delegate != null) {
            delegate.updateAllMarkers();
        }
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/RegionBlueMapLayer.kt
package org.oddlama.vane.regions

import org.bukkit.plugin.Plugin
import org.oddlama.vane.annotation.config.ConfigBoolean
import org.oddlama.vane.annotation.config.ConfigDouble
import org.oddlama.vane.annotation.config.ConfigInt
import org.oddlama.vane.annotation.lang.LangMessage
import org.oddlama.vane.core.lang.TranslatedMessage
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.core.module.ModuleComponent
import org.oddlama.vane.regions.region.Region
import java.util.*

/**
 * BlueMap integration component for rendering region boundaries.
 */
class RegionBlueMapLayer(context: Context<Regions?>) :
    ModuleComponent<Regions?>(context.group("BlueMap", "Enable BlueMap integration.")) {
    /**
     * Owning regions module instance.
     */
    private val regions: Regions
        get() = requireNotNull(module)

    @ConfigBoolean(def = false, desc = "If the marker set should be hidden by default.")
    /**
     * Whether the BlueMap marker set is hidden by default.
     */
    var configHideByDefault: Boolean = false

    @ConfigBoolean(
        def = true,
        desc = "Set to false to make the area markers visible through terrain and other objects."
    )
    /**
     * Whether marker depth testing is enabled.
     */
    var configDepthTest: Boolean = false

    @ConfigInt(def = 2, min = 1, desc = "Area marker line width.")
    /**
     * Marker outline width.
     */
    var configLineWidth: Int = 0

    @ConfigInt(def = 0xffb422, min = 0, max = 0xffffff, desc = "Area marker fill color (0xRRGGBB).")
    /**
     * Marker fill color (RGB integer).
     */
    var configFillColor: Int = 0

    @ConfigDouble(def = 0.1, min = 0.0, max = 1.0, desc = "Area marker fill opacity.")
    /**
     * Marker fill opacity.
     */
    var configFillOpacity: Double = 0.0

    @ConfigInt(def = 0xffb422, min = 0, max = 0xffffff, desc = "Area marker line color (0xRRGGBB).")
    /**
     * Marker outline color (RGB integer).
     */
    var configLineColor: Int = 0

    @ConfigDouble(def = 1.0, min = 0.0, max = 1.0, desc = "Area marker line opacity.")
    /**
     * Marker outline opacity.
     */
    var configLineOpacity: Double = 0.0

    @LangMessage
    /**
     * Localized label for the BlueMap marker set.
     */
    var langLayerLabel: TranslatedMessage? = null

    @LangMessage
    /**
     * Localized marker label format string.
     */
    var langMarkerLabel: TranslatedMessage? = null

    /**
     * Runtime delegate handling BlueMap API calls.
     */
    private var delegate: RegionBlueMapLayerDelegate? = null

    /**
     * Enables BlueMap integration after plugin lookup succeeds.
     */
    fun delayedOnEnable() {
        val plugin: Plugin = regions.server.pluginManager.getPlugin("BlueMap") ?: return

        delegate = RegionBlueMapLayerDelegate(this)
        delegate?.onEnable(plugin)
    }

    /**
     * Schedules delayed BlueMap initialization.
     */
    override fun onEnable() {
        scheduleNextTick { this.delayedOnEnable() }
    }

    /**
     * Disables BlueMap integration and clears delegate state.
     */
    override fun onDisable() {
        delegate?.onDisable()
        delegate = null
    }

    /**
     * Updates one region marker on BlueMap.
     */
    fun updateMarker(region: Region) {
        delegate?.updateMarker(region)
    }

    /**
     * Removes one region marker from BlueMap.
     */
    fun removeMarker(regionId: UUID) {
        delegate?.removeMarker(regionId)
    }

    /**
     * Rebuilds all BlueMap markers from current region state.
     */
    fun updateAllMarkers() {
        delegate?.updateAllMarkers()
    }
}
